staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go

```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
 func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]string, error) {
 	zonedNodes := splitNodesByZone(nodes)
 	klog.V(2).Infof("ensureInternalInstanceGroups(%v): %d nodes over %d zones in region %v", name, len(nodes), len(zonedNodes), g.region)
+
 	var igLinks []string
-	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+	gceZonedNodes := map[string][]string{}
+	for zone, zNodes := range zonedNodes {
+		hosts, err := g.getFoundInstanceByNames(nodeNames(zNodes))
+		if err != nil {
+			return nil, err
+		}
+		names := sets.NewString()
+		for _, h := range hosts {
+			names.Insert(h.Name)
+		}
+		skip := sets.NewString()
+
+		igs, err := g.candidateExternalInstanceGroups(zone)
+		if err != nil {
+			return nil, err
+		}
+		for _, ig := range igs {
+			if strings.EqualFold(ig.Name, name) {
```

Comment:
shouldn't we skip this zone entirely if the name is an exact match, because the name matching means it could be the instance group created with the older style code that includes all the nodes in the zone?

Reply:
> because the name matching means it could be the instance group created with the older style code that includes all the nodes in the zone

There is no guarantee that old ig has all the required nodes.
Also the ig might exist from a previous partial case, where it added some nodes..

This makes sure that we reuse as much as possible, and later on the remaining nodes will be reconciled in k8s managed IG anyways.